src/btree/bt_read.c

```diff
@@ -684,9 +696,16 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
 			 * Check if we need an autocommit transaction.
 			 * Starting a transaction can trigger eviction, so skip
 			 * it if eviction isn't permitted.
+			 *
+			 * The logic here is a little weird: some code paths do
+			 * a blanket ban on checking the cache size in
+			 * sessions, but still require a transaction (e.g.,
+			 * when updating metadata or lookaside).  Make sure we
+			 * start a transaction if the session flag is set.
 			 */
-			return (LF_ISSET(WT_READ_NO_EVICT) ? 0 :
-			    __wt_txn_autocommit_check(session));
+			return (LF_ISSET(WT_READ_IGNORE_CACHE_SIZE) &&
+			    !F_ISSET(session, WT_SESSION_IGNORE_CACHE_SIZE) ?
+			    0 : __wt_txn_autocommit_check(session));
```

Comment:
I think the comment needs more. It took me a while to think through this.  My interpretation ended up something like:
If WT_READ_IGNORE_CACHE_SIZE was passed in explicitly we're done.  If we set WT_READ_IGNORE_CACHE_SIZE because it was set in the session (e.g. ...) then make sure we start a transaction.